src/SAWScript/MGU.hs

```diff
@@ -204,32 +244,57 @@ instance Instantiate Type where
 
 -- }}}
 
--- TI Monad {{{
+
+------------------------------------------------------------
+-- Pass context {{{
+
+--
+-- The monad for this pass is "TI", which is composed of a "readonly"
+-- part (which is not constant or readonly, but where changes are
+-- scoped by the recursive structure of the code) and a read-write
+-- part that accumulates as we move through the code.
+--
+-- XXX: the "readonly" part is used to implement scoping, which is
+-- fine in theory, but in practice because we have declarations that
+-- update the environment, the recursive structure of the code does
+-- not naturally match the scoping. The result is that the recursive
+-- structure of the code has been twisted around to make it work;
+-- that isn't desirable and the organization should probably be
+-- revised.
+--
+-- Anyhow, the elements of the context are:
+--    varEnv: the variable typing environment (variable name to type scheme)
+--    typedefEnv: typedefs (typedef name to its expansion type)
+--    nextTypeIndex: the next fresh unification var number
+--    subst: the unification var substitution we're accumulating
+--    errors: any type errors we've generated so far
```

Comment:
To me those 13 lines really need to be read as a single declaration (and as has come up before, I don't like comments too closely interspersed with code) and the comments don't fit on one line along with the declarations, but it can be done at the cost of the silly comma syntax which I also don't like :-)

done in bf86a7f58c25f50c872b7d3455033fa784d7192e, lmk if you hate it that way though